File: app/src/test/java/com/teamunemployment/lolanalytics/LoginTests/LoginPresenterTests.java
package com.teamunemployment.lolanalytics.LoginTests;

import android.widget.ArrayAdapter;

import com.teamunemployment.lolanalytics.Login.ArrayAdapterFactory;
import com.teamunemployment.lolanalytics.Login.LoginBasePresenter;
import com.teamunemployment.lolanalytics.Login.LoginModel;
import com.teamunemployment.lolanalytics.Login.LoginView;

import org.junit.Test;


import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * @author Josiah Kendall
 */

public class LoginPresenterTests {

    @Test
    public void TestThatWeSetRegionSpinner() {
        ArrayAdapterFactory arrayAdapterFactory = mock(ArrayAdapterFactory.class);
        ArrayAdapter<CharSequence> adapter = mock(ArrayAdapter.class);
        when(arrayAdapterFactory.getArrayAdapter(anyInt(), anyInt())).thenReturn(adapter);
        LoginModel loginModel = mock(LoginModel.class);
        LoginBasePresenter loginPresenter = new LoginBasePresenter(arrayAdapterFactory, loginModel);
        LoginView loginView = mock(LoginView.class);
        loginPresenter.setView(loginView);

        verify(loginView, times(1)).setRegionSpinnerAdapter(any(ArrayAdapter.class));
    }

    @Test
    public void TestThatWeCanNofityUserOfFailedLogin() {

    }

    @Test
    public void TestTHatWeCanLaunchActivityIfSuccessfulLogin() {

    }

    @Test
    public void TEstThatWeSaveUserDetailsGivenCorrectLogin() {

    }


}

File: app/src/test/java/com/teamunemployment/lolanalytics/LoginTests/LoginPresenterTests.kt
package com.teamunemployment.lolanalytics.LoginTests

import android.widget.ArrayAdapter
import com.teamunemployment.lolanalytics.login.sign_in.*
import org.junit.Before
import org.junit.Test
import org.mockito.*
import org.mockito.Mockito.*

/**
 * Created by Josiah Kendall
 */
class LoginPresenterTests {

    @Captor
    private lateinit var arrayAdapterCaptor : ArgumentCaptor<ArrayAdapter<CharSequence>>

    private lateinit var arrayAdapterFactory : ArrayAdapterFactory
    private lateinit var loginInteractor: LoginInteractor
    private lateinit var presenter : LoginPresenter
    private val view = mock(LoginContract.LoginView::class.java)
    private val loginErrorMessages = LoginErrorMessages()

    @Before
    fun `Do The Mahi`() {
        MockitoAnnotations.initMocks(this)
        arrayAdapterFactory = mock(ArrayAdapterFactory::class.java)
        loginInteractor = mock(LoginInteractor::class.java)
        presenter = LoginPresenter(arrayAdapterFactory, loginInteractor)
    }

    @Test
    fun `Make sure that we can handle a 404 when logging in a user`() {

        presenter.setView(view)
        presenter.handleSyncResult(404, 123456)
        verify(view, times(1)).showMessage(loginErrorMessages.`404`())
    }

    @Test
    fun `Make sure that we hide the login button while login is progressing`() {
        presenter.setView(view)
        `when`(view.userName).thenReturn("work cunt")
        `when`(view.region).thenReturn("kotlin is dope")
        presenter.requestSync()
        verify(view, times(1)).hideLoginButton()
    }

    @Test
    fun `Make sure that we show the login button again if we fail`() {
        presenter.setView(view)
        presenter.handleError(Throwable("IoException or some shit"))
        verify(view, times(1)).showLoginButton()
    }

    @Test
    fun `Make sure that we show the login button again after resuming activity`() {
        presenter.setView(view)
        presenter.restart()
        verify(view, times(1)).showLoginButton()
    }

    @Test
    fun `Make sure that we hide login progress on error`() {
        presenter.setView(view)
        presenter.restart()
        verify(view, times(1)).hideProgressSpinner()
    }

    @Test
    fun `Make sure that the progress spinner is shown while we are fetching the user`() {
        presenter.setView(view)
        `when`(view.userName).thenReturn("work cunt")
        `when`(view.region).thenReturn("kotlin is dope")
        presenter.requestSync()
        verify(view, times(1)).showLoginProgressSpinner()
    }

    @Test
    fun `Make sure that the progress spinner is hidden when we restart`() {
        presenter.setView(view)
        presenter.restart()
        verify(view, times(1)).hideProgressSpinner()
    }




}